cpp/include/cuml/experimental/fil/detail/infer_kernel/gpu.cuh

```diff
@@ -0,0 +1,255 @@
+ * expect for non-local categorical data storage.
+ * @param forest The forest used to perform inference
+ * @param postproc The postprocessor object used to store all necessary
+ * data for postprocessing
+ * @param output Pointer to the device-accessible buffer where output
+ * should be written
+ * @param input Pointer to the device-accessible buffer where input should be
+ * read from
+ * @param row_count The number of rows in the input
+ * @param col_count The number of columns per row in the input
+ * @param num_outputs The expected number of output elements per row
+ * @param shared_mem_byte_size The number of bytes of shared memory allocated
+ * to this kernel.
+ * @param output_workspace_size The total number of temporary elements required
+ * to be stored as an intermediate output during inference
+ * @param vector_output_p If non-nullptr, a pointer to the stored leaf
+ * vector outputs for all leaf nodes
+ * @param categorical_data If non-nullptr, a pointer to where non-local
+ * data on categorical splits are stored.
+ */
+template<
+  bool has_categorical_nodes,
+  index_type chunk_size,
+  typename forest_t,
+  typename vector_output_t=std::nullptr_t,
+  typename categorical_data_t=std::nullptr_t
+>
+__global__ void __launch_bounds__(MAX_THREADS_PER_BLOCK, MIN_BLOCKS_PER_SM)
+infer_kernel(
+    forest_t forest,
+    postprocessor<typename forest_t::io_type> postproc,
+    typename forest_t::io_type* output,
+    typename forest_t::io_type const* input,
+    index_type row_count,
+    index_type col_count,
+    index_type num_outputs,
+    index_type shared_mem_byte_size,
+    index_type output_workspace_size,
+    vector_output_t vector_output_p=nullptr,
+    categorical_data_t categorical_data=nullptr
+) {
+  auto constexpr has_vector_leaves = !std::is_same_v<vector_output_t, std::nullptr_t>;
+  auto constexpr has_nonlocal_categories = !std::is_same_v<categorical_data_t, std::nullptr_t>;
+  extern __shared__ std::byte shared_mem_raw[];
+
+  auto shared_mem = shared_memory_buffer(shared_mem_raw, shared_mem_byte_size);
+
+  using node_t = typename forest_t::node_type;
+
+  using output_t = std::conditional_t<
+    has_vector_leaves,
+    std::remove_pointer_t<vector_output_t>,
+    typename node_t::threshold_type
+  >;
+
+  using io_t = typename forest_t::io_type;
+
+  for (
+    auto i=blockIdx.x * chunk_size;
+    i < row_count;
+    i += chunk_size * gridDim.x
+  ) {
+
+    shared_mem.clear();
+    auto* output_workspace = shared_mem.fill<output_t>(output_workspace_size);
+
+    // Handle as many rows as requested per loop or as many rows as are left to
+    // process
+    auto rows_in_this_iteration = min(chunk_size, row_count - i);
+
+    auto* input_data = shared_mem.copy(
+      input + i * col_count,
+      rows_in_this_iteration,
+      col_count
+    );
+
+    auto task_count = chunk_size * forest.tree_count();
+
+    auto num_grove = raft_proto::ceildiv(
```

Comment:
```suggestion
    auto grove_size = raft_proto::ceildiv(
```
`num_grove` sounds like it's counting the number of groves, when we are actually counting the number of trees in each grove.

Reply:
No, that actually is the number of groves. Walking it through, the typical case is that we have fewer threads per block than tasks. So if we divide the number of threads per block by the chunk size, that tells us how many different trees are working simultaneously on that chunk. Each of those trees belongs to the same grove as the trees which are assigned to that block in the next iteration, so that number of trees is equivalent to the number of groves.

Reply:
Got it. Let's keep `num_grove` then.